source/common/upstream/load_balancer_impl.cc

```diff
@@ -27,13 +27,19 @@ LoadBalancerBase::LoadBalancerBase(const HostSet& host_set, const HostSet* local
         [this](const std::vector<HostSharedPtr>&, const std::vector<HostSharedPtr>&) -> void {
           regenerateZoneRoutingStructures();
         });
-    local_host_set_->addMemberUpdateCb(
+    local_host_set_member_update_cb_handle_ = local_host_set_->addMemberUpdateCb(
         [this](const std::vector<HostSharedPtr>&, const std::vector<HostSharedPtr>&) -> void {
           regenerateZoneRoutingStructures();
         });
   }
 }
 
+LoadBalancerBase::~LoadBalancerBase() {
+  if (local_host_set_member_update_cb_handle_ != nullptr) {
+    local_host_set_->removeMemberUpdateCb(local_host_set_member_update_cb_handle_);
```

Comment:
My first attempt at fixing this was RAII where the handles cleaned up during destruction. This lead into a bunch of other issues that I won't go into, but it turned very ugly during shutdown due to lack of ordering (basically universally unregistering all callbacks is tough with the current code). I can definitely do something like add a remove() method on the handle, and hold that in an RAII holder. I can probably also fix it to all work with 100% RAII cleanup, but that will be a larger change. My plan was to try to fix this in a series of follow ups. I'm happy to try to do it all in this change, but all of the TODOs are kind of related and the change will be very large. Probably better to do it in pieces.